rust/appencryption/examples/custom_metastore.rs

```diff
@@ -0,0 +1,245 @@
+    Metastore,
+};
+use async_trait::async_trait;
+use securememory::protected_memory::DefaultSecretFactory;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+/// Example demonstrating how to implement a custom Metastore for Asherah.
+///
+/// This example shows:
+/// 1. How to create a custom Metastore implementation
+/// 2. How to integrate the custom Metastore with Asherah
+/// 3. Basic encrypt/decrypt operations using the custom Metastore
+
+/// A simple Redis-like Metastore implementation.
+///
+/// This is a simplified example that stores keys in memory with Redis-like TTL features.
+/// In a real application, this would connect to a Redis server.
+struct RedisMetastore {
+    /// In-memory storage map for keys
+    store: RwLock<HashMap<String, (EnvelopeKeyRecord, i64, Option<i64>)>>,
+}
+
+impl RedisMetastore {
+    /// Create a new RedisMetastore
+    fn new() -> Self {
+        Self {
+            store: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Generate a key string from id and created timestamp
+    fn generate_key(id: &str, created: i64) -> String {
+        format!("{}_{}", id, created)
+    }
+
+    /// Set a TTL (time-to-live) for a key in seconds
+    async fn expire(&self, id: &str, created: i64, ttl_seconds: i64) -> bool {
+        let key = Self::generate_key(id, created);
+        let mut store = self.store.write().unwrap();
+
+        if let Some(entry) = store.get_mut(&key) {
+            // Calculate expiration time
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+            let expire_at = now + ttl_seconds;
+
+            // Update the TTL
+            entry.2 = Some(expire_at);
+            return true;
+        }
+
+        false
+    }
+
+    /// Delete a key
+    async fn delete(&self, id: &str, created: i64) -> bool {
+        let key = Self::generate_key(id, created);
+        let mut store = self.store.write().unwrap();
+        store.remove(&key).is_some()
+    }
+
+    /// Check if a key is expired
+    fn is_expired(&self, expire_at: Option<i64>) -> bool {
+        if let Some(expire_time) = expire_at {
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+            return now >= expire_time;
+        }
+        false
+    }
+}
+
+#[async_trait]
+impl Metastore for RedisMetastore {
+    async fn load(
+        &self,
+        id: &str,
+        created: i64,
+    ) -> appencryption::Result<Option<EnvelopeKeyRecord>> {
+        let key = Self::generate_key(id, created);
+        let store = self.store.read().unwrap();
+
+        if let Some((record, _, expire_at)) = store.get(&key) {
+            // Check if the key is expired
+            if self.is_expired(*expire_at) {
+                return Ok(None);
```

Comment:
This function never seems to bubble a real `Error` as result. Should this either return a `Result` or `Option`, but not both wrapped? Converting between the two is possible and but ultimately depends on what the user expectations are in case of missing keys not existing/being expired.